examples/CSharp/Tutorial 3.2 - Diffuse Lighting/Program.cs

```diff
@@ -0,0 +1,226 @@
+            LampShader = new Shader(Gl, "shader.vert", "shader.frag");
+
+            //Start a camera at position 3 on the Z axis, looking at position -1 on the Z axis
+            Camera = new Camera(Vector3.UnitZ * 6, Vector3.UnitZ * -1, Vector3.UnitY, Width / Height);
+        }
+
+        private static unsafe void OnUpdate(double deltaTime)
+        {
+            var moveSpeed = 2.5f * (float)deltaTime;
+
+            if (primaryKeyboard.IsKeyPressed(Key.W))
+            {
+                //Move forwards
+                Camera.Position += moveSpeed * Camera.Front;
+            }
+            if (primaryKeyboard.IsKeyPressed(Key.S))
+            {
+                //Move backwards
+                Camera.Position -= moveSpeed * Camera.Front;
+            }
+            if (primaryKeyboard.IsKeyPressed(Key.A))
+            {
+                //Move left
+                Camera.Position -= Vector3.Normalize(Vector3.Cross(Camera.Front, Camera.Up)) * moveSpeed;
+            }
+            if (primaryKeyboard.IsKeyPressed(Key.D))
+            {
+                //Move right
+                Camera.Position += Vector3.Normalize(Vector3.Cross(Camera.Front, Camera.Up)) * moveSpeed;
+            }
+        }
+
+        private static unsafe void OnRender(double deltaTime)
+        {
+            Gl.Enable(EnableCap.DepthTest);
+            Gl.Clear((uint)(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit));
+
+            VaoCube.Bind();
+            LightingShader.Use();
+
+            //Slightly rotate the cube to give it an angled face to look at
+            LightingShader.SetUniform("uModel", Matrix4x4.CreateRotationY(MathHelper.DegreesToRadians(25f)));
+            LightingShader.SetUniform("uView", Camera.GetViewMatrix());
+            LightingShader.SetUniform("uProjection", Camera.GetProjectionMatrix());
+            LightingShader.SetUniform("objectColor", new Vector3(1.0f, 0.5f, 0.31f));
+            LightingShader.SetUniform("lightColor", Vector3.One);
+            LightingShader.SetUniform("lightPos", LampPosition);
+
+            //We're drawing with just vertices and no indicies, and it takes 36 verticies to have a six-sided textured cube
+            Gl.DrawArrays(PrimitiveType.Triangles, 0, 36);
+
+            LampShader.Use();
+
+            //The Lamp cube is going to be a scaled down version of the normal cubes verticies moved to a different screen location
+            var lampMatrix = Matrix4x4.Identity;
+            lampMatrix *= Matrix4x4.CreateScale(0.2f);
+            lampMatrix *= Matrix4x4.CreateTranslation(LampPosition);
+
+            LampShader.SetUniform("uModel", lampMatrix);
+            LampShader.SetUniform("uView", Camera.GetViewMatrix());
+            LampShader.SetUniform("uProjection", Camera.GetProjectionMatrix());
+
+            Gl.DrawArrays(PrimitiveType.Triangles, 0, 36);
```

Comment:
I don't know GL all that well, but what's going on down here?
I'd suggest you add a comment. (or is this a duplicate?)

Reply:
There are two cubes in the scene, each drawn with a different fragment shader (as one cube is being lit while the other cube just represents the light source). I will make this clearer in the code.